### PR TITLE
Add error description for E0174

### PR DESCRIPTION
**Pull request by cristianoliveira**
Reference for issue: #32777

r? @GuillaumeGomez

Hey Guillaume, sorry for taking too long to do it. I got some unexpected work during the week.

Waiting for your review :)

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -1972,6 +1972,88 @@ To learn more about traits, take a look at the Book:
 https://doc.rust-lang.org/book/traits.html
 "##,
 
+E0174: r##"
+This error occurs because of unboxed closure explicitly methods calls is
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: A bit badly formulated.

File: src/librustc_typeck/diagnostics.rs [hunk 2]
```diff
@@ -1972,6 +1972,88 @@ To learn more about traits, take a look at the Book:
...
+experimental in current Rust version.
+
+Example of erroneous code:
+
+```compile_fail
+ fn foo<F: Fn(&str)>(mut f: F) {
+     f.call(("call",));
+     //~ ERROR explicit use of unboxed closure method `call`
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: You can remove the "~" (for all "//~").

Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Also, please change "ERROR" to "error:".

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -1972,6 +1972,88 @@ To learn more about traits, take a look at the Book:
...
+}
+```
+
+Rust’s implementation of closures is a bit different than other languages.
+They are effectively syntax sugar for traits `Fn`, `FnMut` and `FnOnce`.
+To understand better how the closures are implemented see here:
+https://doc.rust-lang.org/book/closures.html#closure-implementation
+
+To fix it you can just call them using `()`, like `foo()`. Although, when you
+execute the closure with `()` under the hood you are executing the methods
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Please add a comma after "with `()`".

File: src/librustc_typeck/diagnostics.rs [hunk 4]
```diff
@@ -1972,6 +1972,88 @@ To learn more about traits, take a look at the Book:
...
+`call`, `call_mut` or `call_once` calling them explicitly is currently
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: And this is badly formulated from this line.

File: src/librustc_typeck/diagnostics.rs [hunk 5]
```diff
@@ -1972,6 +1972,88 @@ To learn more about traits, take a look at the Book:
...
+experimental.
+
+Example of an implicit call:
+
+```
+fn foo<F: Fn()>(f: F) {
+    f("using ()"); //~ Calling using () it works!
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Remove "~".

**Comment by GuillaumeGomez** [comment]
> Hey Guillaume, sorry for taking too long to do it. I got some unexpected work during the week.

No problem! :smiley: 

So there are some sentences that are badly formulated and a few nits. Nothing big. Thanks for your PR!

File: src/librustc_typeck/diagnostics.rs (hunk 1 above)
Comment (cristianoliveira) [inline src/librustc_typeck/diagnostics.rs]: I'm going change to the original short description error:  "explicit use of unboxed closure methods are experimental"

File: src/librustc_typeck/diagnostics.rs (hunk 2 above)
Comment (cristianoliveira) [inline src/librustc_typeck/diagnostics.rs]: I got those examples from here:
https://github.com/rust-lang/rust/blob/master/src/test/compile-fail/feature-gate-unboxed-closures-method-calls.rs#L14

I'm going to change

File: src/librustc_typeck/diagnostics.rs (hunk 4 above)
Comment (cristianoliveira) [inline src/librustc_typeck/diagnostics.rs]: Yeah... When I read it again after did the PR I thought: "I am missing something there"

How about this:

> To fix it you can just call them using `()`, like `foo()`. When you execute the closure with `()`, under the hood you are executing the method `call`, `call_mut` or `call_once`. However, using them explicitly is currently an experimental feature.

**Comment by cristianoliveira** [comment]
Done. :)

File: src/librustc_typeck/diagnostics.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: A bit more explicit isn't a problem. It's just badly formulated like I said.

File: src/librustc_typeck/diagnostics.rs [hunk 6]
```diff
@@ -1972,6 +1972,88 @@ To learn more about traits, take a look at the Book:
...
 "##,
 
+E0174: r##"
+This error occurs because of the explicit use of unboxed closure methods
+that is an experimental feature in current Rust version.
+
+Example of erroneous code:
+
+```compile_fail
+ fn foo<F: Fn(&str)>(mut f: F) {
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: I just noticed: why this functions' lines have an extra blank space at the beggining?

Comment (cristianoliveira) [inline src/librustc_typeck/diagnostics.rs]: Ops

**Comment by GuillaumeGomez** [comment]
Seems good for me but I'd prefer to have a second opinion on this. Thanks for your work! :)

cc @steveklabnik 

PS: don't forget to squash your commits.

**Comment by cristianoliveira** [comment]
For sure. Let's see what @steveklabnik says so I squash them.

File: src/librustc_typeck/diagnostics.rs [hunk 7]
```diff
@@ -1972,6 +1972,88 @@ To learn more about traits, take a look at the Book:
 https://doc.rust-lang.org/book/traits.html
 "##,
 
+E0174: r##"
+This error occurs because of the explicit use of unboxed closure methods
+that is an experimental feature in current Rust version.
```
Comment (steveklabnik) [inline src/librustc_typeck/diagnostics.rs]: "that are an experimental feature in this version of Rust."

File: src/librustc_typeck/diagnostics.rs [hunk 8]
```diff
@@ -1972,6 +1972,88 @@ To learn more about traits, take a look at the Book:
...
+    foo(bar);
+}
+```
+
+Rust’s implementation of closures is a bit different than other languages.
+They are effectively syntax sugar for traits `Fn`, `FnMut` and `FnOnce`.
+To understand better how the closures are implemented see here:
+https://doc.rust-lang.org/book/closures.html#closure-implementation
+
+To fix it you can just call them using `()`, like `foo()`. When you execute
```
Comment (steveklabnik) [inline src/librustc_typeck/diagnostics.rs]: "just" should probably not be used in documentation or error messages. It doesn't add anything, and can seem negative.

I would also maybe slightly re-phrase: "To fix this you can call them using parenthesis, like this: `foo()`.

File: src/librustc_typeck/diagnostics.rs [hunk 9]
```diff
@@ -1972,6 +1972,88 @@ To learn more about traits, take a look at the Book:
...
+the closure with `()`, under the hood you are executing the method call,
+call_mut or call_once. However, using them explicitly is currently an
```
Comment (steveklabnik) [inline src/librustc_typeck/diagnostics.rs]: These should also be in backticks, and have ()s.

File: src/librustc_typeck/diagnostics.rs [hunk 10]
```diff
@@ -1972,6 +1972,88 @@ To learn more about traits, take a look at the Book:
...
+}
+
+fn main() {
+    foo(bar);
+}
+```
+
+To enable the explicit calls you need to add: `#![feature(unboxed_closures)]`
+
+This feature is still unstable so also you need to add: `#![feature(fn_traits)]`
```
Comment (steveklabnik) [inline src/librustc_typeck/diagnostics.rs]: "so you will also need to add"

File: src/librustc_typeck/diagnostics.rs (hunk 7 above)
Comment (cristianoliveira) [inline src/librustc_typeck/diagnostics.rs]: I got it!

**Comment by cristianoliveira** [comment]
Done.

File: src/librustc_typeck/diagnostics.rs [hunk 11]
```diff
@@ -1972,6 +1972,89 @@ To learn more about traits, take a look at the Book:
...
+fn bar(name: &str) {
+    println!("Calling {} it works!", name);
+}
+
+fn main() {
+    foo(bar);
+}
+```
+
+Rust’s implementation of closures is a bit different than other languages.
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: **’** seems weird for me. **'** wouldn't be better in here? cc @steveklabnik

File: src/librustc_typeck/diagnostics.rs [hunk 12]
```diff
@@ -1972,6 +1972,89 @@ To learn more about traits, take a look at the Book:
...
+fn bar(name: &str) {
+    println!("Calling {} it works!", name);
+}
+
+fn main() {
+    foo(bar);
+}
+```
+
+To enable the explicit calls you need to add: `#![feature(unboxed_closures)]`
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Missing a dot at the end of the sentence.

File: src/librustc_typeck/diagnostics.rs [hunk 13]
```diff
@@ -1972,6 +1972,89 @@ To learn more about traits, take a look at the Book:
...
+
+This feature is still unstable so you will also need to add:
+`#![feature(fn_traits)]`
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Missing a dot at the end of the sentence.

**Comment by GuillaumeGomez** [comment]
Just a few more nits and it'll be good to go!

File: src/librustc_typeck/diagnostics.rs (hunk 11 above)
Comment (cristianoliveira) [inline src/librustc_typeck/diagnostics.rs]: It comes from the documentation take a look:
https://doc.rust-lang.org/book/closures.html#closure-implementation

Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: I'm not sure it applies here.

Comment (cristianoliveira) [inline src/librustc_typeck/diagnostics.rs]: I see... There is another case and they have used ' let me change it.

https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/diagnostics.rs#L1548

**Comment by GuillaumeGomez** [comment]
All good, thanks!

@bors: r+ rollup

File: src/librustc_typeck/diagnostics.rs [hunk 14]
```diff
@@ -1972,6 +1972,89 @@ To learn more about traits, take a look at the Book:
...
+
+To fix this you can call them using parenthesis, like this: `foo()`.
+When you execute the closure with parenthesis, under the hood you are executing
+the method `call`, `call_mut` or `call_once`. However, using them explicitly is
+currently an experimental feature.
+
+Example of an implicit call:
+
+```
+fn foo<F: Fn()>(f: F) {
```
Comment (eddyb) [inline src/librustc_typeck/diagnostics.rs]: Missing argument type to `Fn`.

Comment (cristianoliveira) [inline src/librustc_typeck/diagnostics.rs]: Let me fix it.

**Comment by cristianoliveira** [comment]
Fixed

**Pull request merged**